core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java

```diff
@@ -0,0 +1,580 @@
+                List<Type> types,
+                PartitionFunctionFactory partitionFunctionFactory,
+                PartitioningHandle partitioning,
+                List<Integer> partitionChannels,
+                List<Type> partitionChannelTypes,
+                int partitionCount)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.columnChannels = requireNonNull(columnChannels, "columnChannels is null");
+            this.pageSinkManager = requireNonNull(pageSinkManager, "pageSinkManager is null");
+            requireNonNull(writerTarget, "writerTarget is null");
+            checkArgument(
+                    writerTarget instanceof CreateTarget
+                            || writerTarget instanceof InsertTarget
+                            || writerTarget instanceof RefreshMaterializedViewTarget
+                            || writerTarget instanceof TableExecuteTarget,
+                    "writerTarget must be CreateTarget, InsertTarget, RefreshMaterializedViewTarget or TableExecuteTarget");
+            this.target = writerTarget;
+            this.session = session;
+            this.statisticsAggregationOperatorFactory = requireNonNull(statisticsAggregationOperatorFactory, "statisticsAggregationOperatorFactory is null");
+            this.types = requireNonNull(types, "types is null");
+            this.partitionFunctionFactory = requireNonNull(partitionFunctionFactory, "partitionFunctionFactory is null");
+            this.partitioning = requireNonNull(partitioning, "partitioning is null");
+            this.partitionChannels = requireNonNull(partitionChannels, "partitionChannels is null");
+            this.partitionChannelTypes = requireNonNull(partitionChannelTypes, "partitionChannelTypes is null");
+            this.partitionCount = partitionCount;
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext context = driverContext.addOperatorContext(operatorId, planNodeId, PartitionedTableWriterOperator.class.getSimpleName());
+            Operator statisticsAggregationOperator = statisticsAggregationOperatorFactory.createOperator(driverContext);
+            boolean statisticsCpuTimerEnabled = !(statisticsAggregationOperator instanceof DevNullOperator) && isStatisticsCpuTimerEnabled(session);
+
+            PartitionFunction partitionFunction = partitionFunctionFactory.create(
+                    session,
+                    partitioning,
+                    partitionChannels,
+                    partitionChannelTypes,
+                    Optional.empty(),
+                    partitionCount);
+
+            Function<Page, Page> partitionPagePreparer;
+            if (isSystemPartitioning(partitioning)) {
+                partitionPagePreparer = identity();
```

Comment:
add a comment when it can be a system partitioned (aka preferred insert (arbitrary) partitioning)